src/Workspaces/CoreTest/SolutionTests/SolutionTests.cs

```diff
@@ -459,8 +481,459 @@ public void WithAnalyzerConfigDocumentTextLoader()
             Assert.Throws<ArgumentNullException>(() => solution.WithAnalyzerConfigDocumentTextLoader(null!, loader, PreservationMode.PreserveIdentity));
             Assert.Throws<InvalidOperationException>(() => solution.WithAnalyzerConfigDocumentTextLoader(s_unrelatedDocumentId, loader, PreservationMode.PreserveIdentity));
         }
+
+        [Fact]
+        public void WithProjectAssemblyName()
+        {
+            var projectId = ProjectId.CreateNewId();
+
+            var solution = CreateSolution()
+                .AddProject(projectId, "proj1", "proj1.dll", LanguageNames.CSharp);
+
+            // any character is allowed
+            var assemblyName = "\0<>a/b/*.dll";
+
+            var newSolution = solution.WithProjectAssemblyName(projectId, assemblyName);
+            Assert.Equal(assemblyName, newSolution.GetProject(projectId)!.AssemblyName);
+
+            Assert.Same(newSolution, newSolution.WithProjectAssemblyName(projectId, assemblyName));
+
+            Assert.Throws<ArgumentNullException>("assemblyName", () => solution.WithProjectAssemblyName(projectId, null!));
+            Assert.Throws<ArgumentNullException>("projectId", () => solution.WithProjectAssemblyName(null!, "x.dll"));
+            Assert.Throws<InvalidOperationException>(() => solution.WithProjectAssemblyName(ProjectId.CreateNewId(), "x.dll"));
+        }
+
+        [Fact]
+        public void WithProjectOutputFilePath()
```

Comment:
Consolidate this and the below one? Just extract a common method that takes a `Func<Solution, string, Solution>` that calls the right one?

Reply:
Oh or I see you have SolutionTestHelpers.TestProperty, isn't that what this is for?

Reply:
Yes, can use SolutionTestHelpers.TestProperty - it doesn't save much, but I'll use it.
In general I prefer simple flow and clarity over avoiding duplication of test code.

Reply:
I'm usually OK with duplicating test code as well, up until the point the test has bugs. :smile: